examples/src/main/java/com/example/CommonsTextFuzzerHooks.java

```diff
@@ -0,0 +1,28 @@
+// Copyright 2022 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.example;
+
+import com.code_intelligence.jazzer.api.HookType;
+import com.code_intelligence.jazzer.api.MethodHook;
+import java.lang.invoke.MethodHandle;
+
+public class CommonsTextFuzzerHooks {
+    @MethodHook(type = HookType.REPLACE, targetClassName = "org.apache.commons.text.lookup.StringLookupFactory", targetMethod = "constantStringLookup")
+    public static Object hookConstStringLookup(MethodHandle handle, Object thisObject, Object[] args, int hookId) {
+        // Remove the ConstantStringLookup singleton to avoid the ReflectiveCall
```

Comment:
Could you explain in a bit more detail why this is needed and why this additional hook makes it work? Could `--disabled_hooks`, which can be used to disable a hook by class name, be added to `fuzzer_args` instead?

Reply:
Yeah, I might even add this as a comment to the hook, but the reason is the following.

Without the hooks, the fuzzer would always first discover the following vector:

```
$\373const:jaz.Zer.\322\"}\346}\012\200\177\177\177\177
```

this is a sample inpit the fuzzer finds, but it is enough to be

```
${const:jaz.Zer.e}
```

for it to trigger. This is because the `org.apache.commons.text.lookup.ConstantStringLookup#lookup` method loads a class by name in order to look up a static final member of the class, hence here user input can control class loading and your `ReflectiveCall` sanitizer catches it, the data flow is about this:

https://github.com/apache/commons-text/blob/rel/commons-text-1.9/src/main/java/org/apache/commons/text/lookup/ConstantStringLookup.java#L122 --> 
https://github.com/apache/commons-text/blob/rel/commons-text-1.9/src/main/java/org/apache/commons/text/lookup/ConstantStringLookup.java#L146 --> 
https://github.com/apache/commons-lang/blob/rel/commons-lang-3.11/src/main/java/org/apache/commons/lang3/ClassUtils.java#L1118 --> 
https://github.com/apache/commons-lang/blob/rel/commons-lang-3.11/src/main/java/org/apache/commons/lang3/ClassUtils.java#L1135 --> 
https://github.com/apache/commons-lang/blob/rel/commons-lang-3.11/src/main/java/org/apache/commons/lang3/ClassUtils.java#L1069 --> 
https://github.com/CodeIntelligenceTesting/jazzer/blob/5d91fc6f21490932fca1081b52e3e57e71a3d5fd/sanitizers/src/main/java/com/code_intelligence/jazzer/sanitizers/ReflectiveCall.kt

I wanted to avoid this for the demo / test because I think the script execution sink is a lot more interesting: with this one, the most you can do is read some constant from a class on the classpath if you also get back the substituted string. So to avoid this, I replaced the factory method responsible for creating `ConstantStringLookup` with an implementation that returns `null`. That way I disable the `${const:something}` interpolation so the fuzzer can never run into the above flow.

Reply:
We changed the behavior of `jaz.Zer` so that it only reports an issue if an instance is created. This should remove the need for the extra hook.

Reply:
I would have tried to do that after a rebase, but after resolving conflicts, I see weird bazel-related errors popping up in unrelated parts. I also tried ot just build the current main with bazelisk both on macOS and an Ubuntu virual machine with no success.

The error on macOS:

```
bazelisk clean && bazelisk build //... && bazelisk test //... && bazelisk run --local_cpu_resources=8 //examples:CommonsTextFuzzer
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
INFO: Multiplexer process for Javac has closed its output stream
INFO: Repository com_github_jetbrains_kotlin instantiated at:
  /Users/gyorgydemarcsek/projects/jazzer/WORKSPACE.bazel:10:12: in <toplevel>
  /Users/gyorgydemarcsek/projects/jazzer/init.bzl:27:24: in jazzer_init
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/io_bazel_rules_kotlin/src/main/starlark/core/repositories/initialize.bzl:48:31: in kotlin_repositories
Repository rule kotlin_compiler_repository defined at:
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/io_bazel_rules_kotlin/src/main/starlark/core/repositories/compiler.bzl:22:45: in <toplevel>
INFO: Repository remotejdk11_macos instantiated at:
  /DEFAULT.WORKSPACE.SUFFIX:109:6: in <toplevel>
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/bazel_tools/tools/jdk/remote_java_repository.bzl:48:17: in remote_java_repository
Repository rule http_archive defined at:
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
INFO: Repository jazzer_libfuzzer instantiated at:
  /Users/gyorgydemarcsek/projects/jazzer/WORKSPACE.bazel:6:20: in <toplevel>
  /Users/gyorgydemarcsek/projects/jazzer/repositories.bzl:177:10: in jazzer_dependencies
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
Repository rule http_archive defined at:
  /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
ERROR: /private/var/tmp/_bazel_gyorgydemarcsek/071a3ae2fb95e15eb05d82d6a2a2dd28/external/fmeum_rules_jni/jni/tools/libjvm_stub/BUILD.bazel:72:21: While resolving toolchains for target @fmeum_rules_jni//jni/tools/libjvm_stub:current_java_runtime: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'
If platforms or toolchains are a new concept for you, we'd encourage reading https://bazel.build/concepts/platforms-intro.
ERROR: Analysis of target '//:jazzer' failed; build aborted:
INFO: Elapsed time: 2.278s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (214 packages loaded, 1808 targets\
 configured)
    Fetching repository @local_config_cc; Building xcode-locator
    Fetching https://github.com/.../v1.7.22/kotlin-compiler-1.7.22.zip
    Fetching repository @jazzer_jacoco; starting
    Fetching ...d82d6a2a2dd28/external/jazzer_jacoco; Extracting v0.8.8.tar.gz
    Fetching https://mirror.bazel.build/....19-ca-jdk11.0.15-macosx_x64.tar.gz
```

Ubuntu 22.04:

```
bazelisk clean && bazelisk build //...
INFO: Starting clean (this may take a while). Consider using --async if the clean takes more than several minutes.
ERROR: /home/gdemarcsek/jazzer/src/main/java/com/code_intelligence/jazzer/runtime/BUILD.bazel:59:8: no such target '@local_jdk//:bin/jar': target 'bin/jar' not declared in package '' defined by /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/local_jdk/BUILD.bazel (Tip: use `query "@local_jdk//:*"` to see all the targets in that package) and referenced by '//src/main/java/com/code_intelligence/jazzer/runtime:jazzer_bootstrap_shading_test'
INFO: Repository remote_coverage_tools instantiated at:
  /DEFAULT.WORKSPACE.SUFFIX:4:6: in <toplevel>
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
Repository rule http_archive defined at:
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
INFO: Repository remote_java_tools instantiated at:
  /DEFAULT.WORKSPACE.SUFFIX:363:6: in <toplevel>
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
Repository rule http_archive defined at:
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
INFO: Repository com_github_jetbrains_kotlin instantiated at:
  /home/gdemarcsek/jazzer/WORKSPACE.bazel:10:12: in <toplevel>
  /home/gdemarcsek/jazzer/init.bzl:27:24: in jazzer_init
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/io_bazel_rules_kotlin/src/main/starlark/core/repositories/initialize.bzl:48:31: in kotlin_repositories
Repository rule kotlin_compiler_repository defined at:
  /home/gdemarcsek/.cache/bazel/_bazel_gdemarcsek/b7804ce787ab40bb5bfdfe307e852b25/external/io_bazel_rules_kotlin/src/main/starlark/core/repositories/compiler.bzl:22:45: in <toplevel>
ERROR: Analysis of target '//src/main/java/com/code_intelligence/jazzer/runtime:jazzer_bootstrap_shading_test' failed; build aborted: Analysis failed
INFO: Elapsed time: 1.010s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (121 packages loaded, 83 targets configured)
    currently loading: @rules_pkg//pkg/private ... (4 packages)
    Fetching repository @rules_python; starting
    Fetching ...rcsek/b7804ce787ab40bb5bfdfe307e852b25/external/rules_python; Extracting 0.16.1.tar.gz
    Fetching https://mirror.bazel.build/.../releases/coverage_output_generator-v2.6.zip
    Fetching https://mirror.bazel.build/bazel_java_tools/releases/java/v11.9/java_tools-v11.9.zip
    Fetching https://github.com/JetBrains/kotlin/releases/download/v1.7.22/kotlin-compiler-1.7.22.zip

```

I have no clue so far what to do with this.

Reply:
The first one, `No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type`, indicates that Bazel can not pick up the C/C++ toolchain. Is XCode installed?

The second one is similar, but can not find a local JDK.

Reply:
Could you take another look at the build issue?

Reply:
Yep, it seems a full XCode installation and switching to XCode from the CLI tools solves this issue. I guess in the meantime having the cli tools became insufficient to build the main branch. for some reason.